File: java/src/main/java/com/tiagoamp/gameoflife/model/Universe.java
package com.tiagoamp.gameoflife.model;

import java.util.Arrays;

import static com.tiagoamp.gameoflife.model.State.LIVE;
import static com.tiagoamp.gameoflife.rules.Rules.anyDeadCellWithThreeLiveNeighborsBecomesLiveCell;
import static com.tiagoamp.gameoflife.rules.Rules.anyLiveCellWithTwoOrThreeLiveNeighborsSurvives;

/*
 * The universe of the Game of Life is an infinite, two-dimensional orthogonal grid of square cells.
 */
public class Universe {

    private State[][] grid;
    private final int gridSize;
    private final NeighbourFinder finder;


    public Universe(final State[][] seed) {
        grid = seed;
        gridSize = grid.length;
        finder = new NeighbourFinder();
    }


    public int[][] updateToNextGeneration() {
        grid = calculateNextGeneration();
        return getBinaryGrid();
    };


    private State[][] calculateNextGeneration() {
        State[][] newGeneration = new State[gridSize][gridSize];
        for (int r = 0; r < gridSize; r++) {
            for (int c = 0; c < gridSize; c++) {
                State[] neighbours = finder.getNeighboursStateFrom(r,c);
                State prevState = grid[r][c];
                State newState = (prevState == LIVE) ?
                        anyLiveCellWithTwoOrThreeLiveNeighborsSurvives.apply(neighbours) :
                        anyDeadCellWithThreeLiveNeighborsBecomesLiveCell.apply(neighbours);
                newGeneration[r][c] = newState;
            }
        }
        return newGeneration;
    };

    private int[][] getBinaryGrid() {
        int[][] intGrid = new int[gridSize][gridSize];
        for (int r = 0; r < gridSize; r++) {
            for (int c = 0; c < gridSize; c++) {
                State state = grid[r][c];
                intGrid[r][c] = state.getValue();
            }
        }
        return intGrid;
    }


    class NeighbourFinder {

        /* Every cell interacts with its eight neighbours */
        protected State[] getNeighboursStateFrom(int row, int column) {
            State[] neighbours = new State[8];
            neighbours[0] = getStateFrom(row-1,column-1);
            neighbours[1] = getStateFrom(row-1, column);
            neighbours[2] = getStateFrom(row-1, column+1);
            neighbours[3] = getStateFrom(row,column-1);
            neighbours[4] = getStateFrom(row,column+1);
            neighbours[5] = getStateFrom(row+1,column-1);
            neighbours[6] = getStateFrom(row+1,column);
            neighbours[7] = getStateFrom(row+1,column+1);
            return neighbours;
        }

        private State getStateFrom(int row, int column) {
            int r = rollUpIfNecessary(row);
            int c = rollUpIfNecessary(column);
            return grid[r][c];
        }

        private int rollUpIfNecessary(int value) {
            if (value == -1) return gridSize-1;
            else if (value == gridSize) return 0;
            return value;
        }

    }


    protected NeighbourFinder getFinder() {
        return finder;
    }

}

File: javascript/Universe.js
const rules = require('./rules');

/*
 * The universe of the Game of Life is an infinite, two-dimensional orthogonal grid of square cells.
 */
class Universe {

    constructor(seed) {
        this.grid = seed;
        this.gridSize = seed.length;
    }

    updateToNextGeneration() {
        this.grid = this._calculateNextGeneration();
        return this.grid;
    };


    _calculateNextGeneration() {
        const newGeneration = [];
        for (let r = 0; r < this.gridSize; r++) {
            const row = [];
            for (let c = 0; c < this.gridSize; c++) {
                const neighbours = this.getNeighboursStateFrom(r,c);
                const prevState = this.grid[r][c];
                const newState = (prevState == 1) ?
                        rules.anyLiveCellWithTwoOrThreeLiveNeighborsSurvives(neighbours) :
                        rules.anyDeadCellWithThreeLiveNeighborsBecomesLiveCell(neighbours);
                row[c] = newState;
            }
            newGeneration.push(row);
        }
        return newGeneration;
    };

    /* Every cell interacts with its eight neighbours */
    getNeighboursStateFrom(row,column) {
        const neighbours = [];
        neighbours[0] = this._getStateFrom(row-1,column-1);
        neighbours[1] = this._getStateFrom(row-1, column);
        neighbours[2] = this._getStateFrom(row-1, column+1);
        neighbours[3] = this._getStateFrom(row,column-1);
        neighbours[4] = this._getStateFrom(row,column+1);
        neighbours[5] = this._getStateFrom(row+1,column-1);
        neighbours[6] = this._getStateFrom(row+1,column);
        neighbours[7] = this._getStateFrom(row+1,column+1);
        return neighbours;
    }

    _getStateFrom(row, column) {
        const r = this._rollUpIfNecessary(row);
        const c = this._rollUpIfNecessary(column);
        return this.grid[r][c];
    }

    _rollUpIfNecessary(value) {
        if (value == -1) return this.gridSize-1;
        else if (value == this.gridSize) return 0;
        return value;
    }

}

module.exports = Universe;